codeforamerica/vita-min
# GYR1-506  - Limit Greeter access to clients

## Link to pivotal/JIRA issue
https://codeforamerica.atlassian.net/browse/GYR1-506
## Is PM acceptance required? (delete one)
- Yes - don't merge until JIRA issue is accepted!

## What was done?
The Greeter role had it's `:manage` permission for clients taken away and more granular `:edit` and `:read` permissions were given instead. Now the permissions are granted conditional to the properties of the client the greeter is trying to view.

## How to test?
**NOTE:** There are many permissions and many custom pages for many roles. While the scope of this is only to reduce permissions for one role, correcting is most secure by building up permissions rather than tearing down permissions. The unit tests check for manage for all roles, so existing roles should be unscathed but custom actions may be broken for greeters. Check thoroughly. See the self-review for more details.

### Prerequisites

Ideally, there will be at least three clients.

1. A client attached to an organization **with** `allow_greeters: true` **with** a tax return in one of the whitelisted states (for example `intake_ready`)
2. A client attached to an organization **with** `allow_greeters: true` **without** a tax return in a whitelisted state
3. A client attached to an organization **without** `allow_greeters: true`.

### Testing clients with allowed_greeters orgs

1. Visit `/en/hub/clients` as a greeter
2. Verify that the filter drop down has only the four whitelisted states:
     1. Ready for Review
     2. Greeter - info requested
     3. Needs doc help
     4. Not filing
3. Verify that the clients shown are only ones that are in a whitelisted state
4. Verify that attempting to visit a clients page (`/en/hub/clients/:id`) directly (by looking up the id) when they are not in the correct state 403s.
5. Verify that visiting a clients page when they are in the correct state, works
6. Verify that changing the state via `edit_take_action` works.

### Testing clients without allowed_greeters orgs

No clients from these groups should show up on the index page or any of the detail pages (where it will give a 403).

## Screenshots (for visual changes)
- Before
![Screenshot 2024-07-26 at 11 28 34 AM](https://github.com/user-attachments/assets/ed1be2be-0394-4aa2-89df-018075ce0119)

- After
![Screenshot 2024-07-26 at 11 27 00 AM](https://github.com/user-attachments/assets/eee75886-16a4-4177-a13d-c9de8313b600)
